### ISSUE TITLE
Migration payments.0009_alter_subscription_user fails on upgrade to v3.1

### ISSUE THREAD

**Issue by GermanCoding** [healthchecks/healthchecks#927]
I've recently tried to upgrade a healthchecks instance to v3.1 (from v3.0.1), but I haven't been able to do so succesfully.

My upgrade procedure fails at `./manage.py migrate`, which tries to apply new v3.1 migrations. Specifically, the following migration fails:

```
System check identified some issues:

WARNINGS:
api.Check: (models.W037) MariaDB does not support indexes with conditions.
        HINT: Conditions will be ignored. Silence this warning if you don't care about it.
api.Flip: (models.W037) MariaDB does not support indexes with conditions.
        HINT: Conditions will be ignored. Silence this warning if you don't care about it.
Operations to perform:
  Apply all migrations: accounts, admin, api, auth, contenttypes, logs, payments, sessions
Running migrations:
  Applying payments.0009_alter_subscription_user...Traceback (most recent call last):
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/backends/mysql/base.py", line 75, in execute
    return self.cursor.execute(query, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/healthchecks/venv/lib/python3.11/site-packages/MySQLdb/cursors.py", line 179, in execute
    res = self._query(mogrified_query)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/healthchecks/venv/lib/python3.11/site-packages/MySQLdb/cursors.py", line 330, in _query
    db.query(q)
  File "/var/healthchecks/venv/lib/python3.11/site-packages/MySQLdb/connections.py", line 255, in query
    _mysql.connection.query(self, query)
MySQLdb.DataError: (1265, "Data truncated for column 'user_id' at row 1")

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/healthchecks/healthchecks/./manage.py", line 17, in <module>
    raise e
  File "/var/healthchecks/healthchecks/./manage.py", line 13, in <module>
    execute_from_command_line(sys.argv)
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/core/management/base.py", line 412, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/core/management/base.py", line 458, in execute
    output = self.handle(*args, **options)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/core/management/base.py", line 106, in wrapper
    res = handle_func(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/core/management/commands/migrate.py", line 356, in handle
    post_migrate_state = executor.migrate(
                         ^^^^^^^^^^^^^^^^^
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/migrations/executor.py", line 135, in migrate
    state = self._migrate_all_forwards(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/migrations/executor.py", line 167, in _migrate_all_forwards
    state = self.apply_migration(
            ^^^^^^^^^^^^^^^^^^^^^
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/migrations/executor.py", line 252, in apply_migration
    state = migration.apply(state, schema_editor)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/migrations/migration.py", line 132, in apply
    operation.database_forwards(
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/migrations/operations/fields.py", line 235, in database_forwards
    schema_editor.alter_field(from_model, from_field, to_field)
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/backends/base/schema.py", line 877, in alter_field
    self._alter_field(
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/backends/base/schema.py", line 1134, in _alter_field
    self.execute(
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/backends/base/schema.py", line 201, in execute
    cursor.execute(sql, params)
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 79, in execute
    return self._execute_with_wrappers(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 92, in _execute_with_wrappers
    return executor(sql, params, many, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 100, in _execute
    with self.db.wrap_database_errors:
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/healthchecks/venv/lib/python3.11/site-packages/django/db/backends/mysql/base.py", line 75, in execute
    return self.cursor.execute(query, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/healthchecks/venv/lib/python3.11/site-packages/MySQLdb/cursors.py", line 179, in execute
    res = self._query(mogrified_query)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/healthchecks/venv/lib/python3.11/site-packages/MySQLdb/cursors.py", line 330, in _query
    db.query(q)
  File "/var/healthchecks/venv/lib/python3.11/site-packages/MySQLdb/connections.py", line 255, in query
    _mysql.connection.query(self, query)
django.db.utils.DataError: (1265, "Data truncated for column 'user_id' at row 1")
````

I've played around with the migration and noticed the following. When I delete the migration manually, and then regenerate it (using `./manage.py makemigrations` I get the following prompt:

````
./manage.py makemigrations
It is impossible to change a nullable field 'user' on subscription to non-nullable without providing a default. This is because the database needs something to populate existing rows.
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Ignore for now. Existing rows that contain NULL values will have to be handled manually, for example with a RunPython or RunSQL operation.
 3) Quit and manually define a default value in models.py.
Select an option:
````

If I select option "2" above, I get the same migration as the one in the repository, meaning that applying this migration fails with the exact same error. This makes me inclined to believe that my database may have a NULL value for the user in the table, which is specifically the case that isn't allowed anymore after the migration? I'm not that familiar with the internal database layout, so maybe you know better what's going on here?

**Comment by cuu508** [comment]
Thanks for the report, and the troubleshooting, and sorry for the trouble.

Here's the commit that made the Subscription.user_id non-nullable and added the problematic migration: https://github.com/healthchecks/healthchecks/commit/1b1899052cc47bfa774ee581072414c03f4b3c18

I do remember answering "2" when creating the migration. Before creating the migration I checked around in the code and it looked like there are no cases where a subscription gets created but does not get associated with an user account. I don't remember why the Subscription.user_id field was nullable to start with.

I'm not sure how a subscription with a null user_id ended up in your database. As a guess, perhaps you created a subscription object in the Django admin? 

I'm assuming you are not using payments functionality and so don't actually need the Subscription objects in the database. To get the migration to apply cleanly, delete Subscription objects via the Django admin, or delete all rows from the `payments_subscription` table which would have the same effect.

**Comment by GermanCoding** [comment]
Thanks for getting back to me!

First of all, no worries, I've got backups so all is good. I was just reporting this in case this is a more widespread issue. If it's only me with trouble then I'm sorry for the noise. I did not intend to turn this ticket into a support system 😄.





> I'm not sure how a subscription with a null user_id ended up in your database. As a guess, perhaps you created a subscription object in the Django admin?

I'm not sure either. I don't recall fiddling with the subscription system in the admin panel, but maybe? This instance isn't fresh, so possibly something happened at some point.

> 
> I'm assuming you are not using payments functionality and so don't actually need the Subscription objects in the database. To get the migration to apply cleanly, delete Subscription objects via the Django admin, or delete all rows from the `payments_subscription` table which would have the same effect.

That's correct, I've never used the payments function (or had it enabled in the first place). I looked at the above table in my actual SQL database and found the following:

```
 SELECT * FROM payments_subscription;
+----+-------------+----------------------+-----------------+---------+---------+------------+---------------+-----------+---------------+-------------------+-------------------+------------+
| id | customer_id | payment_method_token | subscription_id | user_id | plan_id | address_id | send_invoices | plan_name | invoice_email | next_billing_date | renew_notice_date | setup_date |
+----+-------------+----------------------+-----------------+---------+---------+------------+---------------+-----------+---------------+-------------------+-------------------+------------+
|  1 |             |                      |                 |    NULL |         |            |             1 |           |               | NULL              | NULL              | NULL       |
+----+-------------+----------------------+-----------------+---------+---------+------------+---------------+-----------+---------------+-------------------+-------------------+------------+
```

This is the only entry in the table. So I'm guessing this is just some junk that somehow ended up there and should just be removed. That's fine for me, thanks for the help!

PS: I deleted the junk row and all is good now.
